File: backend-node/src/services/AlertService.ts
import { broadcastAlert, broadcastToRoom } from "./websocketHub.js";
import {
  getAllAlerts as getAll,
  getAlertsByTouristId,
  getOpenAlerts,
  getAlertById,
  createAlert as createAlertDoc,
  updateAlert as updateAlertDoc,
  createNotification,
  type IAlert,
} from "./mongoStore.js";
import { PoliceDepartmentModel } from "../schemas/index.js";

/** Lightweight alert view for WebSocket and API responses */
interface AlertView {
  id: number;
  touristId: string;
  alertType: string;
  lat?: number | null;
  lng?: number | null;
  status: string;
  message?: string | null;
  createdTime: string;
}

function mapToAlertView(alert: IAlert): AlertView {
  return {
    id: alert.alertId,
    touristId: alert.touristId,
    alertType: alert.alertType,
    lat: alert.latitude,
    lng: alert.longitude,
    status: alert.status,
    message: alert.message,
    createdTime: alert.createdAt ? new Date(alert.createdAt).toISOString() : new Date().toISOString(),
  };
}

export async function createAlert(alert: Partial<IAlert>) {
  // Auto-find nearest station if location provided
  let nearestStationId: string | undefined;
  if (typeof alert.latitude === "number" && typeof alert.longitude === "number") {
    try {
      const nearest = await PoliceDepartmentModel.findOne({
        isActive: true,
        location: {
          $near: {
            $geometry: { type: "Point", coordinates: [alert.longitude, alert.latitude] },
            $maxDistance: 100000, // 100km
          },
        },
      }).lean();
      if (nearest) nearestStationId = nearest._id;
    } catch {
      // Geo index may not be built yet — fall back silently
    }
  }

  const saved = await createAlertDoc({
    ...alert,
    status: alert.status ?? "OPEN",
    nearestStationId,
  });

  if (saved.touristId) {
    await createNotification({
      touristId: saved.touristId,
      title: saved.alertType ?? "Alert",
      message: saved.message ?? "Safety alert received",
      type: "alert",
      sourceTab: "home",
      read: false,
      priority: saved.priority === "CRITICAL" ? "critical" : saved.priority === "HIGH" ? "urgent" : "normal",
      channel: "in_app",
    });
  }

  const view = mapToAlertView(saved);
  broadcastAlert(view);

  // Also push to the specific tourist's room
  if (saved.touristId) {
    broadcastToRoom(`tourist:${saved.touristId}`, { type: "ALERT", payload: view });
  }

  return saved;
}

export async function getActiveAlerts() {
  return getOpenAlerts();
}

export async function getAllAlerts() {
  return getAll();
}

export async function getRecentAlerts(limit: number) {
  const all = await getAll();
  if (limit <= 0 || all.length <= limit) {
    return all;
  }
  return all.slice(0, limit);
}

export async function getAlertsForTourist(touristId: string) {
  return getAlertsByTouristId(touristId);
}

export async function handleSOS(touristId: string, lat?: number, lng?: number, message?: string, media?: string[]) {
  const alertMessage = message ?? `TOURIST IN IMMEDIATE DANGER. LAST LOC: ${lat}, ${lng}`;
  return createAlert({
    touristId,
    latitude: lat,
    longitude: lng,
    alertType: "SOS",
    priority: "CRITICAL",
    message: alertMessage,
    escalationLevel: 3,
    media: media ?? [],
  });
}

/**
 * Silent pre-alert (2-second hold). Creates a PENDING alert for monitoring.
 */
export async function handlePreAlert(touristId: string, lat?: number, lng?: number) {
  return createAlert({
    touristId,
    latitude: lat,
    longitude: lng,
    alertType: "PRE_ALERT",
    priority: "MEDIUM",
    status: "PENDING",
    message: "Silent pre-alert triggered — tourist may need help.",
    preAlertTriggered: true,
    escalationLevel: 1,
  });
}

/**
 * Cancel an active alert (before full dispatch).
 */
export async function cancelAlert(alertId: number): Promise<IAlert | null> {
  const alert = await getAlertById(alertId);
  if (!alert) return null;
  if (alert.status === "RESOLVED" || alert.status === "CANCELLED") return alert;

  const updated = await updateAlertDoc(alertId, {
    status: "CANCELLED",
    cancelledAt: new Date(),
  } as Partial<IAlert>);

  if (updated) {
    broadcastAlert(mapToAlertView(updated));
  }
  return updated;
}

/**
 * Get alert status for polling.
 */
export async function getAlertStatus(alertId: number) {
  const alert = await getAlertById(alertId);
  if (!alert) return null;
  return {
    alertId: alert.alertId,
    status: alert.status,
    priority: alert.priority,
    escalationLevel: alert.escalationLevel,
    nearestStationId: alert.nearestStationId,
    resolvedBy: alert.resolvedBy,
    resolvedAt: alert.resolvedAt,
    createdAt: alert.createdAt,
  };
}

export async function updateAlertStatus(alertId: number, newStatus: string, resolvedBy?: string) {
  const alert = await getAlertById(alertId);
  if (!alert) {
    throw new Error(`Alert not found with ID: ${alertId}`);
  }

  const updateData: Partial<IAlert> = { status: newStatus } as Partial<IAlert>;
  if ((newStatus === "RESOLVED" || newStatus === "DISMISSED") && resolvedBy) {
    (updateData as any).resolvedBy = resolvedBy;
    (updateData as any).resolvedAt = new Date();
    // Compute response time
    if (alert.createdAt) {
      (updateData as any).responseTimeMs = Date.now() - new Date(alert.createdAt).getTime();
    }
  }

  const updated = await updateAlertDoc(alertId, updateData);
  if (updated) {
    broadcastAlert(mapToAlertView(updated));
  }
  return updated;
}

File: backend-spring/src/main/java/com/safarsathi/service/AlertService.java
package com.safarsathi.service;

import com.safarsathi.entity.Alert;
import com.safarsathi.entity.Notification;
import com.safarsathi.repository.AlertRepository;
import com.safarsathi.repository.NotificationRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class AlertService {

    private final AlertRepository alertRepository;
    private final NotificationRepository notificationRepository;
    private final SequenceService sequenceService;
    private final WebSocketService webSocketService;

    /**
     * Create a new alert and broadcast it via WebSocket.
     */
    public Alert createAlert(Alert alert) {
        alert.setAlertId((int) sequenceService.getNextId("alertId"));
        if (alert.getStatus() == null) alert.setStatus("OPEN");
        Alert saved = alertRepository.save(alert);

        // Create a corresponding notification
        if (saved.getTouristId() != null) {
            Notification notification = Notification.builder()
                    .notificationId((int) sequenceService.getNextId("notificationId"))
                    .touristId(saved.getTouristId())
                    .title(saved.getAlertType() != null ? saved.getAlertType() : "Alert")
                    .message(saved.getMessage() != null ? saved.getMessage() : "Safety alert received")
                    .type("alert")
                    .sourceTab("home")
                    .read(false)
                    .build();
            notificationRepository.save(notification);
        }

        webSocketService.broadcastAlert(saved);
        return saved;
    }

    public List<Alert> getActiveAlerts() {
        return alertRepository.findByStatus("OPEN", Sort.by(Sort.Direction.DESC, "createdAt"));
    }

    public List<Alert> getAllAlerts() {
        return alertRepository.findAll(Sort.by(Sort.Direction.DESC, "createdAt"));
    }

    public List<Alert> getRecentAlerts(int limit) {
        List<Alert> all = getAllAlerts();
        if (limit <= 0 || all.size() <= limit) return all;
        return all.subList(0, limit);
    }

    public List<Alert> getAlertsForTourist(String touristId) {
        return alertRepository.findByTouristId(touristId, Sort.by(Sort.Direction.DESC, "createdAt"));
    }

    /**
     * Handle an SOS emergency.
     */
    public Alert handleSOS(String touristId, Double lat, Double lng) {
        String message = String.format("TOURIST IN IMMEDIATE DANGER. LAST LOC: %s, %s", lat, lng);
        Alert alert = Alert.builder()
                .touristId(touristId)
                .latitude(lat)
                .longitude(lng)
                .alertType("SOS")
                .priority("CRITICAL")
                .message(message)
                .build();
        return createAlert(alert);
    }

    /**
     * Update the status of an existing alert.
     */
    public Alert updateAlertStatus(int alertId, String newStatus) {
        Alert alert = alertRepository.findByAlertId(alertId)
                .orElseThrow(() -> new RuntimeException("Alert not found with ID: " + alertId));
        alert.setStatus(newStatus);
        Alert updated = alertRepository.save(alert);
        webSocketService.broadcastAlert(updated);
        return updated;
    }
}
